### ISSUE TITLE
[Feature] Deploy to NCNN support for YOLOX

### ISSUE THREAD

**Issue by sktedro** [open-mmlab/mmdeploy#2021]
### Motivation

I've been trying to deploy my YOLOv8 (MMYOLO) model to NCNN using MMDeploy but could not manage to do it. Trying to deploy MMYOLO's YOLOX to NCNN reveals the same (almost) issue when converting. This could be expected as MMYOLO doesn't seem to support deployment to NCNN, but MMDet does and deploying MMDet's YOLOX to NCNN works.

Briefly:
- mmdet.YOLOX can be deployed to NCNN
- mmyolo.YOLOX cannot be deployed to NCNN because of errors explained later
- mmyolo.YOLOv8 cannot be deployed to NCNN for the same reason as mmyolo.YOLOX (it seems)

I don't yet understand the relationship between MMDetection and MMYOLO to such extent that would make it possible for me to understand why this cannot be done, or fix it.

So I suppose the issue is that MMDeploy doesn't support NCNN backend for MMYOLO models, which is strange because it seems MMDetection and MMYOLO are very similar.

##### Deploy config

```py
onnx_config = dict(
    type='onnx',
    export_params=True,
    keep_initializers_as_inputs=False,
    opset_version=11,
    save_file='end2end.onnx',
    input_names=['input'],
    output_names=['output'],
    input_shape=(640, 640),
    optimize=True)
codebase_config = dict(
    type='mmdet',
    task='ObjectDetection',
    model_type='ncnn_end2end',
    post_processing=dict(
        score_threshold=0.05,
        confidence_threshold=0.005,
        iou_threshold=0.5,
        max_output_boxes_per_class=200,
        pre_top_k=5000,
        keep_top_k=100,
        background_label_id=-1))
backend_config = dict(
    type='ncnn',
    precision='FP32',
    use_vulkan=False)
```

##### Deploy command

Latest pre-trained YOLOX-s from MMDetection was used along with the latest YOLOX-s config:
```sh
export MODEL_CFG=../mmdetection/configs/yolox/yolox_s_8xb8-300e_coco.py
export CHECKPOINT=./mmdet_yolox_s_8x8_300e_coco_20211121_095711-4592a793.pth
python3 ../mmdeploy/tools/deploy.py $DEPLOY_CFG $MODEL_CFG $CHECKPOINT deploy/demo.jpg --work-dir $WORK_DIR --device cpu 2>&1 | tee $WORK_DIR/log
```

##### Deploy log

```
...
04/23 13:42:46 - mmengine - INFO - Finish pipeline mmdeploy.apis.pytorch2onnx.torch2onnx
04/23 13:42:47 - mmengine - INFO - Start pipeline mmdeploy.apis.utils.utils.to_backend in main process
onnx2ncnn finish
04/23 13:42:47 - mmengine - INFO - Finish pipeline mmdeploy.apis.utils.utils.to_backend
04/23 13:42:47 - mmengine - INFO - visualize ncnn model start.
04/23 13:42:50 - mmengine - WARNING - Failed to search registry with scope "mmdet" in the "Codebases" registry tree. As a workaround, the current "Codebases" registry in "mmdeploy" is used to build instance. This may cause unexpected failure when running the built modules. Please check whether "mmdet" is a correct scope, or whether the registry is initialized.
04/23 13:42:50 - mmengine - WARNING - Failed to search registry with scope "mmdet" in the "mmdet_tasks" registry tree. As a workaround, the current "mmdet_tasks" registry in "mmdeploy" is used to build instance. This may cause unexpected failure when running the built modules. Please check whether "mmdet" is a correct scope, or whether the registry is initialized.
04/23 13:42:50 - mmengine - WARNING - Failed to search registry with scope "mmdet" in the "backend_detectors" registry tree. As a workaround, the current "backend_detectors" registry in "mmdeploy" is used to build instance. This may cause unexpected failure when running the built modules. Please check whether "mmdet" is a correct scope, or whether the registry is initialized.
04/23 13:42:51 - mmengine - INFO - visualize ncnn model success.
...
```

##### Deploy config

```py
onnx_config = dict(
    type='onnx',
    export_params=True,
    keep_initializers_as_inputs=False,
    opset_version=11,
    save_file='end2end.onnx',
    input_names=['input'],
    output_names=['dets', 'labels'],
    input_shape=(640, 640),
    optimize=True)

codebase_config = dict(
    type='mmyolo',
    task='ObjectDetection',
    model_type='ncnn_end2end',
    post_processing=dict(
        score_threshold=0.05,
        confidence_threshold=0.005,
        iou_threshold=0.5,
        max_output_boxes_per_class=200,
        pre_top_k=5000,
        keep_top_k=100,
        background_label_id=-1),
        module=['mmyolo.deploy'])
backend_config = dict(
    type='ncnn',
    precision='FP32',
    use_vulkan=False)
```

##### Deploy command

As with MMDetection, latest pre-trained YOLOX-s model with it's config was used:
```sh
export MODEL_CFG=../mmyolo_0.5.0/configs/yolox/yolox_s_fast_8xb8-300e_coco.py
export CHECKPOINT=./mmyolo_0.5.0_yolox_s_fast_8xb8-300e_coco_20230213_142600-2b224d8b.pth
python3 ../mmdeploy/tools/deploy.py $DEPLOY_CFG $MODEL_CFG $CHECKPOINT deploy/demo.jpg --work-dir $WORK_DIR --device cpu 2>&1 | tee $WORK_DIR/log
```

##### Deploy log

```
...
04/23 13:37:47 - mmengine - INFO - Finish pipeline mmdeploy.apis.pytorch2onnx.torch2onnx
04/23 13:37:47 - mmengine - INFO - Start pipeline mmdeploy.apis.utils.utils.to_backend in main process
Unsupported Flatten axis 2!
Unsupported Gather axis: 0
Unknown data type 0
Unknown data type 0
onnx2ncnn finish
04/23 13:37:47 - mmengine - INFO - Finish pipeline mmdeploy.apis.utils.utils.to_backend
04/23 13:37:47 - mmengine - INFO - visualize ncnn model start.
04/23 13:37:50 - mmengine - WARNING - Failed to get codebase, got: 'Cannot get key by value "mmyolo" of <enum \'Codebase\'>'. Then export a new codebase in Codebase MMYOLO: mmyolo
04/23 13:37:51 - mmengine - WARNING - Import mmdeploy.codebase.mmyolo.deploy failedPlease check whether the module is the custom module.No module named 'mmdeploy.codebase.mmyolo'
04/23 13:37:51 - mmengine - WARNING - Failed to search registry with scope "mmyolo" in the "Codebases" registry tree. As a workaround, the current "Codebases" registry in "mmdeploy" is used to build instance. This may cause unexpected failure when running the built modules. Please check whether "mmyolo" is a correct scope, or whether the registry is initialized.
04/23 13:37:51 - mmengine - WARNING - Failed to search registry with scope "mmyolo" in the "backend_detectors" registry tree. As a workaround, the current "backend_detectors" registry in "mmdeploy" is used to build instance. This may cause unexpected failure when running the built modules. Please check whether "mmyolo" is a correct scope, or whether the registry is initialized.
layer TopK not exists or registered
04/23 13:37:51 - mmengine - ERROR - /home/tedro/Desktop/d_projekty/bp/proj/../mmdeploy/tools/deploy.py - create_process - 82 - visualize ncnn model failed.
```

Same output applies to both MMYOLO v0.4.0 and v0.5.0

I also tried changing output_names to `['output']` in the deploy config, but that gave the same output (as expected).

### Deploying MMYOLO.YOLOv8 to NCNN

YOLOv8 model I was trying to deploy was a custom one but pretty much the same as the pre-trained YOLOv8 model in MMYOLO - just the input image size was changed to 640x384 (w*h).

Deploy config used was the same as MMYOLO.YOLOX, deploy commands were pretty much the same but it shouldn't really matter because the output speaks for itself:

##### Deploy log

```
...
04/23 01:59:04 - mmengine - INFO - Finish pipeline mmdeploy.apis.pytorch2onnx.torch2onnx
04/23 01:59:16 - mmengine - INFO - Start pipeline mmdeploy.apis.utils.utils.to_backend in main process
Unsupported Flatten axis 2!
Unsupported Gather axis: 0
Unknown data type 0
Unknown data type 0
onnx2ncnn finish
04/23 02:00:12 - mmengine - INFO - Finish pipeline mmdeploy.apis.utils.utils.to_backend
04/23 02:00:12 - mmengine - INFO - visualize ncnn model start.
04/23 02:00:24 - mmengine - WARNING - Failed to get codebase, got: 'Cannot get key by value "mmyolo" of <enum \'Codebase\'>'. Then export a new codebase in Codebase MMYOLO: mmyolo
04/23 02:00:28 - mmengine - WARNING - Import mmdeploy.codebase.mmyolo.deploy failedPlease check whether the module is the custom module.No module named 'mmdeploy.codebase.mmyolo'
04/23 02:00:28 - mmengine - WARNING - Failed to search registry with scope "mmyolo" in the "Codebases" registry tree. As a workaround, the current "Codebases" registry in "mmdeploy" is used to build instance. This may cause unexpected failure when running the built modules. Please check whether "mmyolo" is a correct scope, or whether the registry is initialized.
04/23 02:00:28 - mmengine - WARNING - Failed to search registry with scope "mmyolo" in the "backend_detectors" registry tree. As a workaround, the current "backend_detectors" registry in "mmdeploy" is used to build instance. This may cause unexpected failure when running the built modules. Please check whether "mmyolo" is a correct scope, or whether the registry is initialized.
layer NonMaxSuppression not exists or registered
04/23 02:00:28 - mmengine - ERROR - deploy/deploy.py - create_process - 45 - visualize ncnn model failed.
...
```

# Summary

MMDeploy is able to deploy a mmdet.YOLOX model to NCNN, but cannot deploy mmyolo.YOLOX model to NCNN. `mmdeploy_onnx2ncnn` gives the same error:
```
Unsupported Flatten axis 2!
Unsupported Gather axis: 0
Unknown data type 0
Unknown data type 0
```
both for mmyolo.YOLOX and mmyolo.YOLOv8 and it seems this error is responsible for the onnx2ncnn outputting a corrupt model (inference doesn't work either - NCNN's input and output names are empty lists so it can't take any input and give any output). 

That makes me think that MMDeploy does generally support exporting YOLO models to NCNN, but doesn't work well with MMYOLO in this use-case.

MMLabs pip versions:
```
mmcv                        2.0.0
mmdeploy                    1.0.0
mmdet                       3.0.0
mmengine                    0.7.2
mmyolo                      0.4.0
```

I can provide more data about my experiments and this issue if needed - onnx or ncnn models, environment variables, full log, precise model and deploy configs used, pip list, .., but I think the data I provided should explain everything.

**Comment by huayuan4396** [comment]
Hi, @sktedro , for yolox, the issue is caused because mmyolo has not rewrite the `YOLOv5Head.predict_by_feat()` for ncnn (yolox is derived from yolov5 at present).
A possible solution is to add the following code in `mmyolo/deploy/models/dense_heads/yolov5_head.py`:
```python
@FUNCTION_REWRITER.register_rewriter(
    func_name='mmyolo.models.dense_heads.yolov5_head.'
    'YOLOv5Head.predict_by_feat',
    backend='ncnn')
def yolov5_head__predict_by_feat__ncnn(self,
                                 cls_scores: List[Tensor],
                                 bbox_preds: List[Tensor],
                                 objectnesses: Optional[List[Tensor]] = None,
                                 batch_img_metas: Optional[List[dict]] = None,
                                 cfg: Optional[ConfigDict] = None,
                                 rescale: bool = False,
                                 with_nms: bool = True) -> Tuple[InstanceData]:
    ctx = FUNCTION_REWRITER.get_context()
    from mmdeploy.codebase.mmdet.ops import ncnn_detection_output_forward
    from mmdeploy.utils import get_root_logger
    from mmdeploy.utils.config_utils import is_dynamic_shape
    dynamic_flag = is_dynamic_shape(ctx.cfg)
    if dynamic_flag:
        logger = get_root_logger()
        logger.warning('YOLOX does not support dynamic shape with ncnn.')
    img_height = int(batch_img_metas[0]['img_shape'][0])
    img_width = int(batch_img_metas[0]['img_shape'][1])

    assert len(cls_scores) == len(bbox_preds) == len(objectnesses)
    device = cls_scores[0].device
    dtype = cls_scores[0].dtype
    cfg = self.test_cfg if cfg is None else cfg
    batch_size = bbox_preds[0].shape[0]
    featmap_sizes = [cls_score.shape[2:] for cls_score in cls_scores]
    mlvl_priors = self.prior_generator.grid_priors(
        featmap_sizes, dtype=dtype, device=device, with_stride=True)
    mlvl_priors = [mlvl_prior.unsqueeze(0) for mlvl_prior in mlvl_priors]
    flatten_priors = torch.cat(mlvl_priors, dim=1)
    flatten_cls_scores = [
        cls_score.permute(0, 2, 3, 1).reshape(batch_size, -1,
                                              self.num_classes)
        for cls_score in cls_scores
    ]
    flatten_bbox_preds = [
        bbox_pred.permute(0, 2, 3, 1).reshape(batch_size, -1, 4)
        for bbox_pred in bbox_preds
    ]
    flatten_objectness = [
        objectness.permute(0, 2, 3, 1).reshape(batch_size, -1, 1)
        for objectness in objectnesses
    ]

    cls_scores = torch.cat(flatten_cls_scores, dim=1).sigmoid()
    dummy_cls_scores = torch.zeros(
        batch_size, cls_scores.shape[-2], 1, device=cls_scores.device)
    batch_mlvl_scores = torch.cat([dummy_cls_scores, cls_scores], dim=2)
    score_factor = torch.cat(flatten_objectness, dim=1).sigmoid()
    flatten_bbox_preds = torch.cat(flatten_bbox_preds, dim=1)
    assert flatten_priors.shape[-1] == 4, f'yolox needs (B, N, 4) priors, got\
            (B, N, {flatten_priors.shape[-1]})'
    prior_box_x1 = (flatten_priors[:, :, 0:1] - flatten_priors[:, :, 2:3] / 2) \
                   / img_width
    prior_box_y1 = (flatten_priors[:, :, 1:2] - flatten_priors[:, :, 3:4] / 2) \
                   / img_height
    prior_box_x2 = (flatten_priors[:, :, 0:1] + flatten_priors[:, :, 2:3] / 2) \
                   / img_width
    prior_box_y2 = (flatten_priors[:, :, 1:2] + flatten_priors[:, :, 3:4] / 2) \
                   / img_height
    prior_box_ncnn = torch.cat(
        [prior_box_x1, prior_box_y1, prior_box_x2, prior_box_y2], dim=2)

    scores = batch_mlvl_scores.permute(0, 2, 1).unsqueeze(3) * \
             score_factor.permute(0, 2, 1).unsqueeze(3)
    scores = scores.squeeze(3).permute(0, 2, 1)

    batch_mlvl_bboxes = flatten_bbox_preds.reshape(batch_size, 1, -1)
    batch_mlvl_scores = scores.reshape(batch_size, 1, -1)
    batch_mlvl_priors = prior_box_ncnn.reshape(batch_size, 1, -1)
    batch_mlvl_vars = torch.ones_like(batch_mlvl_priors)
    batch_mlvl_priors = torch.cat([batch_mlvl_priors, batch_mlvl_vars], dim=1)
    deploy_cfg = ctx.cfg
    post_params = get_post_processing_params(deploy_cfg)
    iou_threshold = cfg.nms.get('iou_threshold', post_params.iou_threshold)
    score_threshold = cfg.get('score_thr', post_params.score_threshold)
    pre_top_k = post_params.pre_top_k
    keep_top_k = cfg.get('max_per_img', post_params.keep_top_k)

    vars = torch.tensor([1, 1, 1, 1], dtype=torch.float32)
    output__ncnn = ncnn_detection_output_forward(
        batch_mlvl_bboxes, batch_mlvl_scores, batch_mlvl_priors,
        score_threshold, iou_threshold, pre_top_k, keep_top_k,
        self.num_classes + 1,
        vars.cpu().detach().numpy())
    return output__ncnn
```
Besides, you should change your `onnx_config.output_names`from `['dets', 'labels']` to `['outputs']`.
This works for me.

**Comment by sktedro** [comment]
Oh, I see. Thank you very much, I will try it out when possible :)

I suppose you can close this, because it seems the issue lies withing MMYOLO, not MMDeploy.

**Comment by huayuan4396** [comment]
Feel free to reopen it when getting any problem.

**Comment by Petros626** [comment]
@huayuan4396 did you got experience with model rewriting for tensorrt export? Unfortunately not all models are supported.
